app/services/email_checker.rb

```diff
@@ -0,0 +1,89 @@
+class EmailChecker
+  extend Forwardable
+  def_delegators :SendgridApi, :bounced?, :spam_reported?
+
+  def initialize(original_address, override_sendgrid = false)
+    @original_address = original_address
+    @parsed = parse_address(original_address)
+    @override_sendgrid = override_sendgrid
+  end
+
+  def error
```

Comment:
Suggestion to simplify the `error` method.

Firstly set `compute_error` to return a default `:valid` symbol when no delivery problems have been encountered:

```ruby
def compute_error
+    return :unparseable unless parsed
+    return :domain_dot if domain_dot_error?
+    return :malformed unless well_formed_address?
+    return :no_mx_record unless mx_records?
+    unless override_sendgrid?
+      return :spam_reported if spam_reported?(parsed.address)
+      return :bounced if bounced?(parsed.address)
+    end
+    :valid
+  end
```

We can then cache the `compute_error` return value and use active supports inquiry method so we can just delegate the `valid?` predicate method to the instances `:error` method - which now returns a string inquiry object.

```ruby
def error
  @error ||= compute_error.to_s.inquiry
end

delgate :valid?, to: :error
```

Copied the class and updated as a <a href="https://gist.github.com/dan-palmer/bd5ca41c89ba9c878455">gist</a> for copy and paste if you like the change.